src/framework/components/rigid-body/system.js

```diff
@@ -552,6 +601,275 @@ class RigidBodyComponentSystem extends ComponentSystem {
+     * @param {Vec3|Quat} [rotation] - The world space rotation for the shape to have.
+     *
+     * @returns {HitResult[]} An array of shapeTest hit results (0 length if there were no hits).
+     * @private
+     */
+    _shapeTest(shape, position = Vec3.ZERO, rotation = Vec3.ZERO) {
+        Debug.assert(Ammo.ConcreteContactResultCallback, 'pc.RigidBodyComponentSystem#_shapecast: Your version of ammo.js does not expose Ammo.ConcreteContactResultCallback. Update it to latest.');
+
+        const results = [];
+
+        // Set proper position
+        ammoVec3.setValue(position.x, position.y, position.z);
+
+        // Set proper rotation
+        if (rotation instanceof Quat) {
+            ammoQuat.setValue(rotation.x, rotation.y, rotation.z, rotation.w);
+        } else {
+            ammoQuat.setEulerZYX(rotation.z, rotation.y, rotation.x);
+        }
+
+        // Assign position and rotation to transform.
+        ammoTransform.setIdentity();
+        ammoTransform.setOrigin(ammoVec3);
+        ammoTransform.setRotation(ammoQuat);
+
+        // We only initialize the shapeTast body here so we don't have an extra body unless the user uses this function
+        if (!shapeTestBody) {
+            shapeTestBody = this.createBody(0, shape, ammoTransform);
+        }
+
+        // Make sure the body has proper shape, transform and is active.
+        shapeTestBody.setCollisionShape(shape);
+        shapeTestBody.setWorldTransform(ammoTransform);
+        shapeTestBody.forceActivationState(BODYSTATE_ACTIVE_TAG);
+
+        // Callback for the contactTest results.
+        const resultCallback = new Ammo.ConcreteContactResultCallback();
+        resultCallback.addSingleResult = function (cp, colObj0Wrap, partId0, index0, colObj1Wrap, p1, index1) {
+            // Retrieve collided entity.
+            const body1 = Ammo.castObject(Ammo.wrapPointer(colObj1Wrap, Ammo.btCollisionObjectWrapper).getCollisionObject(), Ammo.btRigidBody);
+
+            // Make sure there is an existing entity.
+            if (body1.entity) {
+                // Retrieve manifold point.
+                const manifold = Ammo.wrapPointer(cp, Ammo.btManifoldPoint);
+
+                // Make sure there is a collision
+                if (manifold.getDistance() < 0) {
+                    const point = manifold.get_m_positionWorldOnB();
+                    const normal = manifold.get_m_normalWorldOnB();
+
+                    // Push the result.
+                    results.push(new HitResult(
+                        body1.entity,
+                        new Vec3(point.x(), point.y(), point.z()),
+                        new Vec3(normal.x(), normal.y(), normal.z())
+                    ));
+                }
+            }
+        };
+
+        // Check for contacts.
+        this.dynamicsWorld.contactTest(shapeTestBody, resultCallback);
+
+        // Disable body and remove shape.
+        shapeTestBody.forceActivationState(BODYSTATE_DISABLE_DEACTIVATION);
```

Comment:
Hmm. BODYSTATE_DISABLE_DEACTIVATION is used, because we don't have CF_NO_CONTACT_RESPONSE defined, and they both happen to equal to 4. The result will be the same, but I personally don't like a flag that was created for different purpose is used.

Reply:
Actually we do have it: BODYFLAG_NORESPONSE_OBJECT
Was it the flag you wanted to use? BODYSTATE_DISABLE_DEACTIVATION is used when we want to prevent the body from sleeping.

Reply:
I'm not sure to get it, how would a flag named BODYSTATE_DISABLE_DEACTIVATION would not be to set body state to disabled?

Reply:
Well I was trying to disable the shape, so I should use BODYFLAG_NORESPONSE_OBJECT?

Reply:
Well, to be perfectly honest, I won't be able to tell for sure, unless I tinker with it myself :) Setting a state to BODYSTATE_DISABLE_DEACTIVATION would just prevent it from sleeping, it won't disable it.

You would have to test it, but I think it should go something like:
```
body.setCollisionFlags(body.getCollisionFlags() | BODYFLAG_NORESPONSE_OBJECT);
```

Reply:
Collision flag? Is it really useful to disable collision when body is removed from world and shape is removed?

Reply:
Ah, I might have missed where the body is removed from the world. I thought only the shape was destroyed. If it is, then we don't need that line. It was the comment there `// Disable body and remove shape` that didn't match the code, which didn't disable the body. If the body is removed, that whole line can be removed then altogether. Actually, if the shape is removed, then there should be no collisions reported either.

Reply:
Interesting to know, thanks! I removed the line.